### PR TITLE
Style notifications using onyxData, increase payload bandwidth

### PR DESCRIPTION
**Pull request by Julesssss**
### Details

The notification payload size limit is 4kb and going over this causes the notification to be dropped by FCM. To resolve multiple issues we need to reduce the amount of payload data that is being sent unnecessarily. This PR prepares for this by switching the notification data source from reportComment to the onyxData, allowing us to remove lots of duplicate fields once this PR has been deployed to users.  

### Fixed Issues

$ https://github.com/Expensify/App/issues/14849
$ https://github.com/Expensify/Expensify/issues/269103
$ https://github.com/Expensify/Expensify/issues/260142
$ https://github.com/Expensify/App/issues/15423
[UNBLOCKS] https://github.com/Expensify/App/issues/14715
[UNBLOCKS] https://github.com/Expensify/Expensify/issues/265452

### Tests

_Screenshots for each of these tests are at the bottom of this description._ 

**Setup steps**
- Use [this StackOverflow guide](https://stackoverflowteams.com/c/expensify/questions/15025) to build a debuggable Android release build
    - step 1 is optional
    - step 2 is NOT optional!
- Start your dev environment and NGROK
- Open the app and accept the 'Allow notifications' prompt if using a newish OS version
- Sign into the Android app using one of your dev users
- Feel free to DM me for help!

**1) Verify that messages longer than 115 chars are no longer being truncated**
- From a separate client, send this message to the Android user:
    - this is a very very very very very very very very very very very very very very very very very very very very very very long message that wasn't previously possible! 😮😮😮😮😮😮😮😮
- The full message should be displayed
- The message should not contain a `"` symbol at the end

**2) Verify that messages and 1:1 IOU requests are displayed correctl**y
- From a separate client, send any message to the Android user
- From a separate client, send an IOU request to the Android user for any amount
- Both messages should be displayed correctly (ignoring the currency symbol issue)
- Both messages should be part of the same thread/notification (this only applies to newer Android versions)
- _See the screenshot below for an example_

**3) Verify that bill split messages are displayed correctly**
- From a separate client, create a new group that includes the Android user
- Send any message to the group
- It should be received by the android user
- Create a bill split between the group
- It should be received by the android user and formatted correctly (ignoring the currency symbol issue)
- These will not be threaded, as they are from two separate reports

### Offline tests

N/A -- notifications cannot be received while the device is offline

### QA Steps

Run the above tests, ignoring the setup steps

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos

**Notes:** 
- _ONLY ANDROID NATIVE CODE IS MODIFIED_
- _CURRENCY SYMBOLS NOT RENDERING IS A SEPARATE PRE-EXISTING ISSUE_

**1) Verify that messages longer than 115 chars are no longer being truncated**
Android 8 | Android 13
| - | - |
long-android-8 | long-android-13

**2) Verify that messages and 1:1 IOU requests are displayed correctly**
Android 8 | Android 13
| - | - |
other-android-8 | other-android-13

**3) Verify that bill split messages are displayed correctly**
Android 8 | Android 13
| - | - |
noti-android-8 | noti-android-13

### CODE REVIEW

File: android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java [hunk 1]
```diff
@@ -8,36 +8,40 @@
...
 import android.graphics.Bitmap;
+import android.graphics.Bitmap.Config;
 import android.graphics.Canvas;
 import android.graphics.Paint;
-import android.graphics.Rect;
-import android.graphics.Bitmap.Config;
-import android.graphics.PorterDuffXfermode;
 import android.graphics.PorterDuff.Mode;
+import android.graphics.PorterDuffXfermode;
+import android.graphics.Rect;
```
Comment (Julesssss) [inline android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java:16]: This import reordering was done by automatic code formatting

**Comment by Julesssss** [comment]
@jasperhuangg @parasharrajat @arosiclair @cristipaval 

Ready for testing now. I used physical devices, but I think these steps should work fine on an emulator. DM me if you want me to save you all the setup steps and I'll send you a signed release apk file you can install to any emulator or physical device!

**Review by arosiclair — DISMISSED** [review]

**Comment by jasperhuangg** [comment]
@parasharrajat I'll remove you for review from this since you can't really test this one

**Comment by jasperhuangg** [comment]
@Julesssss 

All the test cases seem to work as expected for me except for test case 3.

It seems the split bill notification appears in the same thread for me as the message that I sent into that same group.
Screenshot 2023-03-17 at 2 46 19 PM

File: android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java [hunk 2]
```diff
@@ -92,12 +94,13 @@ protected NotificationCompat.Builder onExtendBuilder(@NonNull Context context, @
...
         }
 
+        // Attempt to parse data and style notification
         if (message.containsKey(PAYLOAD_KEY)) {
             try {
                 JsonMap payload = JsonValue.parseString(message.getExtra(PAYLOAD_KEY)).optMap();
 
-                // Apply message style only for report comments
-                if (REPORT_COMMENT_TYPE.equals(payload.get(TYPE_KEY).getString())) {
+                // Apply message style when onyxData is present
```
Comment (jasperhuangg) [inline android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java:102]: to clarify, this has to be Onyx data that we receive from Pusher? Urban Airship? just curious

Comment (Julesssss) [reply android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java:102]: For Android/iOS we receive the onyxData update from both Pusher (only when the app is open) and Airship (we get the push notification regardless of whether the app is alive/killed). 

So in this code we only care that the notification data includes onyxData. I'll update the comment to make this a bit clearer.

**Review by cristipaval — DISMISSED** [review]

**Review by cristipaval — APPROVED** [review]

**Comment by Julesssss** [comment]
> All the test cases seem to work as expected for me except for test case 3.
> 
> It seems the split bill notification appears in the same thread for me as the message that I sent into that same group. Could I be missing something?

Hey @jasperhuangg thanks for thorough testing. I should have explained better, but that is expected because (as far as I can tell) you sent that message from the group chat. This is the logic for showing in multiple threads: 

- All my 1:1 dm messages with you should be in a single notification thread
- If I request money from you, that should show in the same thread
- Messaged in a group chat between me you and Gandalf will show in a separate notification thread
- Group split notifications will show in the same group chat thread
- The 1:1 notification will NOT show in the same group chat thread

There are some existing bugs which break the thread logic, but based on what you said I think this is totally expected 👍

**Comment by Julesssss** [comment]
@jasperhuangg please take another look, I think we're good here now. What you saw was expected.

**Comment by jasperhuangg** [comment]
@Julesssss Thanks for your patience was OOO, grabbing screenshots now and will merge when done!

**Comment by jasperhuangg** [comment]
@Julesssss 

Ahh sorry, I can't seem to get any notifications on my android emulator, can someone else help with screenshots?

**Review by jasperhuangg — APPROVED** [review]

**Comment by jasperhuangg** [comment]
@Julesssss 

Ahh sorry, I can't seem to get any notifications on my android emulator and I can't figure out why, can someone else help with screenshots?

cc @arosiclair @cristipaval

**Comment by Julesssss** [comment]
@jasperhuangg I'd ask for help from someone with a physical deice, or to use browser stack (though I haven't tested notifications using that service). It's much easier using a compiled build on a physical device.

**Comment by jasperhuangg** [comment]
Tried a number of things and still am not getting any notifications on my Android emulator

@cristipaval @arosiclair Reached out on NewDot DMs for help with this, let's get this merged!

**Review by arosiclair — APPROVED** [review]
Tests well for me 👍🏽

**Comment by jasperhuangg** [comment]
@Julesssss

> do you remember if you switched the airship keys? That might be the missing step.

I'm pretty sure I did, I took everything you removed from this commit and added it back locally: https://github.com/Expensify/App/pull/15857/commits/3d816d30c494d95596491ba8c4213c537df193b5

**Comment by Julesssss** [comment]
Ah cool, next time we can try debugging together 👍

**Comment by Julesssss** [comment]
Verified on staging ✅ 

Screenshot_20230324-184559
Screenshot_20230324-184628
Screenshot_20230324-184452

**Pull request merged**